### PR TITLE
add sudoku.scrypt

### PR DESCRIPTION
**Pull request by zhfnjust**

### CODE REVIEW

File: contracts/sudoku.scrypt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    function formatCheck(bytes solution) : bool {
+        require(len(solution) == Sudoku.N * Sudoku.N);
+        return true;
+    }
+
+    function merge(bytes solution) : bytes {
+
+        bytes newBoard = this.board;
+        int i = 0;
+        // N - 2
```
Comment (xhliu) [inline contracts/sudoku.scrypt:22]: remove

File: tests/js/sudoku.scrypttest.js [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+];
+
+
+const row = 9;
+const col = 9;
+function boardToHex(arr) {
+  let hex = '';
+  for (let i = 0; i < row; i++) {
+    for (let j = 0; j < col; j++) {
+      hex = hex + (num2bin(arr[i][j]) === ''   ? '00' : num2bin(arr[i][j]));
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:91]: hex = hex + (arr[i][j]) === 0   ? '00' : num2bin(arr[i][j]));

File: tests/js/sudoku.scrypttest.js [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+describe('Test sCrypt contract sudoku In Javascript', () => {
+  let sudoku, preimage, result
+
+  before(() => {
+    const Sudoku = buildContractClass(compileContract('sudoku.scrypt'))
+    sudoku = new Sudoku(new Bytes(boardToHex(board)))
+
+  })
+
+  it('should success', () => {
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:108]: success -> succeed

File: tests/js/sudoku.scrypttest.js [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    result = sudoku.isValidSudoku(new Bytes(boardToHex(solution))).verify()
+    expect(result.success, result.error).to.be.true
+  });
+
+
+  it('should failed', () => {
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:114]: failed -> fail

File: tests/js/sudoku.scrypttest.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    result = sudoku.isValidSudoku(new Bytes(boardToHex(validSolution0))).verify()
+    expect(result.success, result.error).to.be.false
+  });
+
+
+  it('should failed', () => {
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:120]: same

File: tests/js/sudoku.scrypttest.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+  [8,3,6,7,5,2,1,4,9],
+  [0,6,3,2,7,4,9,1,8],
+  [4,8,7,0,3,9,6,2,5],
+  [9,2,1,5,6,8,0,7,4],
+  [6,0,2,3,4,5,8,0,1],
+  [3,9,0,8,2,1,5,6,7],
+  [1,0,8,6,9,7,4,0,2]
+];
+
+const validSolution2 = [
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:39]: Unused. Delete.

File: tests/js/sudoku.scrypttest.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+  [3,9,4,8,2,1,5,6,7],
+  [1,5,8,6,9,7,4,3,2]
+];
+
+
+
+
+
+
+const validSolution0 = [
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:57]: Let's name this invalid, since it is NOT a valid solution.

File: tests/js/sudoku.scrypttest.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  [4,8,7,1,3,9,6,2,5],
+  [9,2,1,5,6,8,3,7,4],
+  [6,7,2,3,4,5,8,9,1],
+  [3,9,4,8,2,1,5,6,7],
+  [1,5,8,6,9,7,4,3,2]
+];
+
+
+// board, 0 is empty value
+const board = [
```
Comment (xhliu) [inline tests/js/sudoku.scrypttest.js:26]: Let's have a "sparse" board, i.e., more `0`s, to be realistic.

File: contracts/sudoku.scrypt [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+
+
+
+
+
+
+    public function isValidSudoku(bytes solution) {
+
+        require(this.formatCheck(solution));
```
Comment (xhliu) [inline contracts/sudoku.scrypt:55]: No need for a function here. Just check length.

File: contracts/sudoku.scrypt [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        return newBoard;
+    }
+
+
+
+
+
+
+
+    public function isValidSudoku(bytes solution) {
```
Comment (xhliu) [inline contracts/sudoku.scrypt:53]: isValidSudoku -> solve

Pub function name should be a verb, like `unlock`.

File: contracts/sudoku.scrypt [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+
+                if (value == Sudoku.EMPTY) {
+                    bytes inputValue = Util.getElemAt(solution, this.index(i, j));
+
+                    int elem = Util.fromLEUnsigned(inputValue);
+
+                    require(elem <= 9);
+
+                    newBoard = Util.setElemAt(newBoard, this.index(i, j), inputValue);
+                }
```
Comment (xhliu) [inline contracts/sudoku.scrypt:37]: What's the solution format? Non-zero element in the initial board has to show up as well or not? If yes,
```
 else {
  // check `solution`'s non-empty element is the same with `this.board`, i.e., u cannot change the initial board.
}
```

If not, make sure they are 0.
```
else {
  // check 0
}
```

File: contracts/sudoku.scrypt [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+        loop (9) {
+            int j = 0;
+            int rowSum = 0;
+            int colSum = 0;
+            int squareSum = 0;
+
+            loop (9) {
+
+    
+                int rowElem = Util.fromLEUnsigned(Util.getElemAt(newBord, this.index(i, j)));
```
Comment (xhliu) [inline contracts/sudoku.scrypt:70]: Put this common code into a function

`Util.fromLEUnsigned(Util.getElemAt(newBord, this.index(i, j)))`

File: contracts/sudoku.scrypt [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+
+
+
+
+
+    public function isValidSudoku(bytes solution) {
+
+        require(this.formatCheck(solution));
+
+        bytes newBord = this.merge(solution);
```
Comment (xhliu) [inline contracts/sudoku.scrypt:57]: merge -> formatCheck

File: contracts/sudoku.scrypt [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+                colSum = colSum + colElem;
+                int squareElem = Util.fromLEUnsigned(Util.getElemAt(newBord, this.indexSquare(i, j)));
+                squareSum = squareSum + squareElem;
+
+                j = j + 1;
+            }
+
+            require(rowSum == 45);
+            require(colSum == 45);
+            require(squareSum == 45);
```
Comment (xhliu) [inline contracts/sudoku.scrypt:83]: Is this a sufficient condition to ensure a solution is found? Definitely a necessary condition. Can u have duplicate digits in this case, e.g., every element is set to 5, given initial board of only `0`s and `5`s. U need to check row, col, and subgrid has unique digits.

**Review by xhliu — CHANGES_REQUESTED** [review]
🥇

File: tests/js/sudoku.scrypttest.js (hunk 8 above)
Comment (xhliu) [reply tests/js/sudoku.scrypttest.js:26]: E.g.
https://en.wikipedia.org/wiki/Sudoku#Variations_of_grid_sizes_or_region_shapes
https://rosettacode.org/wiki/Category:BBC_BASIC

File: contracts/sudoku.scrypt (hunk 13 above)
Comment (zhfnjust) [reply contracts/sudoku.scrypt:41]: it not only format checking.

File: contracts/array.scrypt [hunk 15]
```diff
@@ -0,0 +1,127 @@
+
+contract Array {
```
Comment (xhliu) [inline contracts/array.scrypt:2]: 👍 

I was planning to do this. We can wrap it into the standard library later for dynamic array.

**Review by xhliu — APPROVED** [review]
awesome 💯

**Pull request merged**